File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <map>

using namespace std;

int n, m;
int sum[501];
bool visited[501][501];
vector<vector<int>> board;
map<int, int> colList;

int dy[4] = {0, 1, 0, -1};
int dx[4] = {1, 0, -1, 0};

bool inRange(int y, int x) {
    return 0 <= y && y < n && 0 <= x && x < m;
}

void bfs() {
    queue<pair<int, int>> q;
    
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
           if(board[i][j] == 0 || visited[i][j]) continue;
                    
            q.push({i, j});
            visited[i][j] = true;
            int cnt = 1;
            //map<int, int> colList;
            colList.clear();
            colList[j] = j;
            
            while(!q.empty()) {
                pair<int, int> cur = q.front();
                q.pop();
                
                int y = cur.first, x = cur.second;
                for(int dir = 0; dir < 4; dir++) {
                    int ny = y + dy[dir], nx = x + dx[dir];
                    if(!inRange(ny, nx) || visited[ny][nx] || !board[ny][nx]) continue;
                    cnt++;
                    visited[ny][nx] = true;
                    if(colList.find(nx) == colList.end())
                        colList[nx] = nx;
                    q.push({ny, nx});
                }
            }
            
            for(auto iter = colList.begin(); iter != colList.end(); iter++) {
                sum[iter->first] += cnt; 
            }
        }
    } 
}

int solution(vector<vector<int>> land) {
    int answer = 0;
    
    n = land.size();
    m = land[0].size();
    board = land;
    
    bfs();
    
    for(int i = 0; i < m; i++) {
        answer = max(answer, sum[i]);
    }
        
    return answer;
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;

class Solution {
    
    int n, m;
    int[][] land;
    int[] res;
    boolean[][] visited;
    
    int[] dy = {0, 1, 0, -1};
    int[] dx = {1, 0, -1, 0};
    
    public boolean inRange(int y, int x) {
        return 0 <= y && y < n && 0 <= x && x < m;
    }
    
    public void bfs(int r, int c) {
        
        Queue<int[]> q = new LinkedList<>();
        Set<Integer> set = new HashSet<>();
       	
        set.add(c);
        q.offer(new int[]{r, c});
       	visited[r][c] = true; 
        
        int cnt = 1;
        while(!q.isEmpty()) {
            int[] here = q.poll();
            int y = here[0], x = here[1];
            
            for(int dir = 0; dir < 4; dir++) {
                int ny = y + dy[dir], nx = x + dx[dir];
               	if(!inRange(ny, nx) || land[ny][nx] == 0 || visited[ny][nx]) continue;
                set.add(nx);
                visited[ny][nx] = true;
               	cnt++;
                q.offer(new int[]{ny, nx});
            }
        } 
        
        for(int col : set) {
         	res[col] += cnt;
        }
    }
    
    public void calcSum() {
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(!visited[i][j] && land[i][j] == 1)
                	bfs(i, j);
            }
        }
    }
    
    public int solution(int[][] land) {
       	this.land = land; 
       	n = land.length;
        m = land[0].length;
        visited = new boolean[n][m];
        res = new int[m];
        
        calcSum();
        int answer = 0;
        for(int i = 0; i < m; i++) {
            answer = Math.max(answer, res[i]);
        }
        return answer;
    }
}